Framework/CurveFitting/src/Functions/CubicSpline.cpp

```diff
@@ -85,25 +79,39 @@ void CubicSpline::setupInput(boost::scoped_array<double> &x,
     std::string yName = "y" + num;
 
     x[i] = getAttribute(xName).asDouble();
+    y[i] = getParameter(yName);
 
-    // if x[i] is out of order with its neighbours
-    if (i > 1 && i < n && (x[i - 1] < x[i - 2] || x[i - 1] > x[i])) {
-      xSortFlag = true;
+    if (!xSortFlag) {
+      // if x[i] is out of order with its neighbours
+      if (i > 1 && i < n && (x[i - 1] < x[i - 2] || x[i - 1] > x[i])) {
+        xSortFlag = true;
+      }
     }
-
-    y[i] = getParameter(yName);
   }
 
   // sort the data points if necessary
   if (xSortFlag) {
     g_log.warning() << "Spline x parameters are not in ascending order. Values "
                        "will be sorted.\n";
-    std::sort(x.get(), x.get() + n);
+
+    using point = std::pair<double, double>;
+    std::vector<point> pairs;
```

Comment:
The sorting looks good now.

My only remaining comment would be that as we know the size of the final pairs vector it would be good to set that initially, either in the constructor or using `.reserve`, to avoid `push_back` having to resize the vector as it goes.